### PR TITLE
Update Main.sol

### PR DESCRIPTION
**Pull request by asuzukosi**

### CODE REVIEW

File: smart-contract/contracts/Main.sol [hunk 1]
```diff
@@ -230,32 +230,125 @@ contract ZuriSchoolVoting {
...
+    function setWeight(string memory stakeholder, uint256 weight) public isChairperson {
+        require(weight >= 0, "weights can not be less than 1");
+        if(stringsEquals(stakeholder,"student")){
+            weights[Stakeholder.STUDENT] = weight;
+        } else if(stringsEquals(stakeholder,"teacher")){
+            weights[Stakeholder.TEACHER] = weight;
+        } else if(stringsEquals(stakeholder,"director")){
+            weights[Stakeholder.DIRECTOR]  = weight;
+        }else {
+            require(false, "invalid stakeholder name entered");
```
Comment (punk1234) [inline smart-contract/contracts/Main.sol:327]: What do you think about using `revert` here ?

File: smart-contract/contracts/Main.sol [hunk 2]
```diff
@@ -230,32 +230,125 @@ contract ZuriSchoolVoting {
...
 
 
-    // @function used to customize the vote weight for the various stakeholders todo @Kosi
+    /**
+     * @notice view the results of a completed election
+     * @dev view results of an election
+     * @param _electionId the id of the election you want to view the results for
+     */
+    function viewResult(uint256 _electionId) public view returns(string memory electionName, string memory proposalName, uint256 voteCount){
+        require(elections[_electionId].computed == true, "results have not yet been compiled");
```
Comment (punk1234) [inline smart-contract/contracts/Main.sol:307]:
```suggestion
        require(elections[_electionId].computed, "results have not yet been compiled");
```

File: smart-contract/contracts/Main.sol [hunk 3]
```diff
@@ -230,32 +230,125 @@ contract ZuriSchoolVoting {
...
+     /**
+     * @notice compile the results of an election, the election is automatically stopped when the result is compiled
+     * @dev compile election results
+     * @param _electionId the id of the election you want to compile the results for
+     */
+    function compileResults(uint256 _electionId) public isDirectorOrTeacher {
+        stopElection(_electionId);
+        elections[_electionId].computed = true;
+
+        Proposal memory _max = choices[_electionId][0];
```
Comment (punk1234) [inline smart-contract/contracts/Main.sol:289]:
```suggestion
        Proposal memory _leadingCandidate = choices[_electionId][0];
```

File: smart-contract/contracts/Main.sol [hunk 4]
```diff
@@ -230,32 +230,125 @@ contract ZuriSchoolVoting {
...
-    // @function used to compile results of an election. should only be called by a director or teacher todo @Kosi
+    /**
+     * @notice view the stats of an ongoing or completed election
+     * @dev view election statistics
+     * @param _electionId the id of the election you want to view the statistics of
+     */
+    function viewElectionStats(uint256 _electionId) public view returns(string[] memory names, string[] memory user_types, bool[] memory canVotes, bool[] memory hasVoteds){
+  
+        string[] memory _names = new string[](voter_addresses.length);
+        string[] memory _user_types = new string[](voter_addresses.length);
```
Comment (punk1234) [inline smart-contract/contracts/Main.sol:257]: `voter_addresses` is now `voterAddresses` in the contract

**Review by punk1234 — CHANGES_REQUESTED** [review]

File: smart-contract/contracts/Main.sol (hunk 1 above)
Comment (asuzukosi) [reply smart-contract/contracts/Main.sol:327]: I actually don't know how to use revert, but you can merge it and make the changes from your end then ill pull it on my fork

**Pull request merged**